### PR TITLE
refactor product card graphql query

### PR DESCRIPTION
**Pull request by davidleger95**
## Testing

- go to service card in storybook
- does it load correctly?

## Checklist

- [ ] **CHANGELOG**: The **Unreleased** section of CHANGELOG was updated
- [ ] **Prop changes**: [docs][docs] were updated
- [ ] **Prop changes**: E2E tests were updated, testing from the highest level possible
- [ ] **Platform testing**: If this change should be [tested against a platform][platform-testing], has it been?

[docs]: https://ui.sandbox.manifold.co
[platform-testing]: https://app.gitbook.com/@manifold/s/engineering/playbooks/testing-your-code-against-a-platform

### CODE REVIEW

File: src/components/manifold-service-card/manifold-service-card.tsx [hunk 1]
```diff
@@ -1,34 +1,17 @@
...
-            free
-          }
-        }
-      }
-    }
-  }
-`;
+import productCardQuery from './product-card.graphql';
+
+type PartialProduct = Omit<ProductCardQuery, 'product'>;
```
Comment (davidleger95) [inline src/components/manifold-service-card/manifold-service-card.tsx]: This doesn't work for some reason 🤔 any thoughts?

File: src/components/manifold-service-card/manifold-service-card.tsx [hunk 2]
```diff
@@ -1,34 +1,17 @@
...
-            free
-          }
-        }
-      }
-    }
-  }
-`;
+import productCardQuery from './product-card.graphql';
+
+type PartialProduct = Pick<ProductCardQuery, 'product'>;
```
Comment (davidleger95) [inline src/components/manifold-service-card/manifold-service-card.tsx]: This doesn't work for some reason 🤔 any thoughts?

**Comment by davidleger95** [comment]
no sure why the formatting in `src/spec/graphql.schema.json` is all messed up. doesn't `npm run generate:gql` format that for us?

File: src/components/manifold-service-card/manifold-service-card.tsx (hunk 2 above)
Comment (dangodev) [reply src/components/manifold-service-card/manifold-service-card.tsx]: Try `ProductCardQuery['product']` (you can just specify this on the `@Prop`)

File: src/components/manifold-service-card/product-card.graphql [hunk 3]
```diff
@@ -0,0 +1,17 @@
+query ProductCard($productLabel: String!) {
+    product(label: $productLabel) {
+      id
+      displayName
+      tagline
+      label
+      logoUrl
+      # must get all plans until 'free' filter works
+      plans(first: 500, orderBy: { field: COST, direction: ASC }) {
```
Comment (dangodev) [inline src/components/manifold-service-card/product-card.graphql]: Is this just to see if a product has 1 free plan? Instead you can do `plans(first: 1, free: true)`. If it returns `1` it has a free plan; if `0` it doesn’t.

This was broken before, but we can use that now.

**Comment by dangodev** [comment]
@davidleger95 Have you done an `npm install` recently? You might have an out-of-date gql-codegen (`graphql.schema.json` shouldn’t have changed that much).

File: src/components/manifold-service-card/product-card.graphql (hunk 3 above)
Comment (davidleger95) [reply src/components/manifold-service-card/product-card.graphql]: so free plans are properly ordered now? would that mean the `free` filter works as well? so we could do this:
```graphql
plans(first: 1, free: true) { ... }
```

Comment (dangodev) [reply src/components/manifold-service-card/product-card.graphql]: Yup I think that’s what I said above

**Comment by davidleger95** [comment]
Github keeps emailing me saying all the checks failed on this PR... but they all passed 🤔

File: src/components/manifold-service-card/product-card.graphql [hunk 4]
```diff
@@ -0,0 +1,16 @@
+query ProductCard($productLabel: String!) {
+  product(label: $productLabel) {
+    id
+    displayName
+    tagline
+    label
+    logoUrl
+    plans(first: 1, free: true) {
```
Comment (dangodev) [inline src/components/manifold-service-card/product-card.graphql]: Alternately, we could do `freePlans: plans(first: 1, free: true)` and change the logic to see if `freePlans.edges.length > 0` to see if it’s free or not.

FWIW I think this is fine, and not a suggestion to change it. Just a micro-optimization I guess.

File: src/components/manifold-service-card/manifold-service-card.tsx [hunk 5]
```diff
@@ -83,9 +68,10 @@ export class ManifoldServiceCard {
       return;
     }
 
-    const { data, errors } = await this.graphqlFetch({
-      query,
-      variables: { productLabel },
+    const variables: ProductCardQueryVariables = { productLabel };
```
Comment (dangodev) [inline src/components/manifold-service-card/manifold-service-card.tsx]: 👍

File: src/types/unpack.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export type Unpack<T, K extends string> = T extends { [k in K]: infer U }
+  ? U // if T has K then we extarct the type of K
+  : T;
```
Comment (dangodev) [inline src/types/unpack.ts]: What’s the advantage of this as opposed to just using `ProductCardQuery['product']`?

File: src/components/manifold-service-card/product-card.graphql (hunk 4 above)
Comment (davidleger95) [reply src/components/manifold-service-card/product-card.graphql]: I think that's what we're doing. We can't leave the query body blank so we have to add at least one field. for example, this:
```graphql
{
	product(label: "scoutapp") {
    plans(first: 1, free: true)
  }
}
```
sadly gives us this error:
```
"Field \"plans\" of type \"PlanConnection\" must have a selection of subfields. Did you mean \"plans { ... }\"?"
```

so `free` seemed like a reasonable field to grab here. 

A while  ago I suggested having a `count` field so we could just do this:

```graphql
{
	product(label: "scoutapp") {
    freePlans: plans(first: 1, free: true) {
      count
    }
  }
}
```

but I don't think that's a high priority for backend at the moment

File: src/types/unpack.ts (hunk 6 above)
Comment (davidleger95) [reply src/types/unpack.ts]: I thoguht I tried that and didn't think it worked. I'll double check. that would be cleaner

File: src/components/manifold-service-card/product-card.graphql (hunk 4 above)
Comment (dangodev) [reply src/components/manifold-service-card/product-card.graphql]: Yeah I’d like a `count` too, but I talked with James and he said it’s actually a design (performance) choice not to have a total count. So we probably won’t ever get it 😞

File: src/types/unpack.ts (hunk 6 above)
Comment (davidleger95) [reply src/types/unpack.ts]: 🤦‍♂ welp, that's good to know that works at least! 😄

**Pull request merged**